docs: reviewed HCL code generation (#967)
terramate-io/terramate: docs/code-generation/generate-hcl.md

  link: '/code-generation/generate-file'
---

> # Generating HCL Code with Terramate

Author: Okay. 

Although a more explicit search like "how to generate HCL code with Terramate" might enhance search engine optimization to direct users here, I fully understand the rationale behind excluding "Terramate" to avoid redundancy.

> ## Configuration files and `generate_hcl` blocks

Reviewer: ```suggestion
## Configuration files and `generate_hcl` blocks
```

Also here inconsistencies upper and lower-case ...

> To begin generating HCL code with Terramate, developers need to define configuration files that specify the desired code components and their corresponding values. Terramate utilizes the `generate_hcl` block to generate HCL code based on these configurations. Multiple generate_hcl blocks can be defined within a [configuration file](https://terramate.io/docs/cli/configuration/), each responsible for generating a specific code component.

Reviewer: ```suggestion
To begin generating HCL code with Terramate, developers need to define configuration files that specify the desired code components and their corresponding values. Terramate utilizes the `generate_hcl` block to generate HCL code based on these configurations. Multiple `generate_hcl` blocks can be defined within a [configuration file](../configuration/index.md), each responsible for generating a specific code component.
```

Reviewer: @Adeniyikayodee please never use direct links but instead let's use the internals of Vitepress to link to other documentation

Reviewer: @Adeniyikayodee we should also limit the line length to 120 max ...

## Understanding the `generate_hcl` block structure

> The `generate_hcl` block follows a specific structure, consisting of a label, a file path, and a content block. The label provides a unique identifier for the generated code component, while the file path determines the location where the generated code will be saved. The content block contains the HCL code that will be generated.

Reviewer 2: ```
The label provides a unique identifier for the generated code component, while the file path 
determines the location where the generated code will be saved.
```
The label **is** the file path, then the snippet above needs some rewording or maybe I didn't understand. Also I don't get what is the "generated code component" in this case here.

To begin generating HCL code with Terramate, developers need to define configuration files that specify the desired code components and their corresponding values. Terramate utilizes the `generate_hcl` block to generate HCL code based on these configurations. Multiple generate_hcl blocks can be defined within a [configuration file](https://terramate.io/docs/cli/configuration/), each responsible for generating a specific code component.

HCL code generation is done using `generate_hcl`
blocks in [Terramate configuration files](../configuration/index.md).
## Understanding the `generate_hcl` block structure

The code may include:
The `generate_hcl` block follows a specific structure, consisting of a label, a file path, and a content block. The label provides a unique identifier for the generated code component, while the file path determines the location where the generated code will be saved. The content block contains the HCL code that will be generated.

* Blocks, sub blocks, etc
* Attributes initialized by literals
* Terramate Global references
* Terramate Metadata references
* Expressions using interpolation, functions, etc

> * Dynamic blocks using the `tm_dynamic` block type

Reviewer 2: Is it already clear that all of this is supported in the `content` block without mentioning them? @mariux 
The `content` block is exceptionally the only block where the user is free to include custom blocks, subblocks etc at will, all the other Terramate blocks have a fixed schema.
Maybe the wording can be changed to make that clear without having to enumerate all of the possible features allowed?

Reviewer 3: yes, this should be kept/made clear...

Author: Okay. I will include this.

> The label is defined using the label key within the `generate_hcl` block. It should be a descriptive name that helps identify the purpose of the generated code component. 

Reviewer 2: From where this come from?
There's no `label` attribute in the generate blocks.

> For example, a label could be `backend_configuration` for generating backend configurations or `provider_configuration` for generating provider configurations.

Reviewer 2: this makes no sense at all ...

> The file path is specified using the `file_path` key within the `generate_hcl` block. It determines the location and file name of the generated code. Developers can provide a relative or absolute path based on their project structure and requirements.

Reviewer 2: wtf? haha

  link: '/code-generation/generate-file'
---

# HCL Code Generation

Terramate supports the generation of arbitrary HCL code referencing Terramate defined data.

The generated code can then be composed/referenced by any Terraform code inside a stack (or any other
tool that uses HCL, like [Packer](https://www.packer.io/)).

HCL code generation is done using `generate_hcl` blocks in Terramate [configuration file](../configuration/index.md).

The code may include:

- Blocks, sub blocks, etc
- Attributes initialized by literals
- Terramate Global references
- Terramate Metadata references
- Expressions using interpolation, functions, etc
- Dynamic blocks using the tm_dynamic block type

> ## Content block for code generation

Reviewer: ```suggestion
## Content block for code generation

```

line breaks missing

  link: '/code-generation/generate-file'
---

# HCL Code Generation

> ## Configuration Files and `generate_hcl` Blocks

Reviewer: ```suggestion
## Configuration Files and `generate_hcl` Blocks

```

To begin generating HCL code, developers need to define configuration files that 
specify the desired code components and their corresponding values. Multiple `generate_hcl` 
blocks can be defined within a [configuration file](../configuration/index.md), 
each responsible for generating a specific code component.

HCL code generation is done using `generate_hcl` blocks in Terramate configuration files.

The code may include:

- Blocks, sub blocks, etc
- Attributes initialized by literals
- Terramate Global references
- Terramate Metadata references
- Expressions using interpolation, functions, etc
- Dynamic blocks using the tm_dynamic block type

> ## Content Block for Code Generation

Reviewer: ```suggestion
## Content Block for Code Generation

```

The content block is where the actual HCL code is defined. It uses the content key within the `generate_hcl` 
block. The content can include HCL code snippets, expressions, interpolations, functions, and references to 
Terramate-defined data.

Developers can leverage the full power of HCL to define the desired code component within the content block. 
This includes using variables, loops, conditionals, and other HCL constructs to dynamically generate code based 
on specific conditions or requirements.

> ## Generating Backend Configurations for Stacks

Reviewer: ```suggestion
## Generating Backend Configurations for Stacks

```

Terramate enables the generation of backend configurations for stacks within a project. 

Backend configurations define where Terraform stores its state file and provide access to remote state storage, 
such as AWS S3, Azure Blob Storage, or HashiCorp Consul.

> ## Example: Generating a Backend Configuration

Reviewer: ```suggestion
## Example: Generating a Backend Configuration

```

  link: '/code-generation/generate-file'
---

# HCL Code Generation

> Terramate supports the generation of arbitrary HCL code referencing Terramate defined data.

Reviewer: ```suggestion
Terramate supports the generation of arbitrary HCL code referencing [Terramate defined data](../data-sharing/index.md).
```

Reviewer: we should not remove any links

}
```

> ## Generating Provider and Terraform Configurations for Stacks

Reviewer: ```suggestion
## Generating Provider and Terraform Configurations for Stacks

```

Terramate also facilitates the generation of provider configurations and the main Terraform configuration for 
stacks within a project. Provider configurations specify the providers to be used and their corresponding 
configuration settings. 

The main Terraform configuration contains the resource and module definitions required to provision the 
infrastructure.

> ## Example: Generating Provider and Terraform Configurations

Reviewer: ```suggestion
## Example: Generating Provider and Terraform Configurations

```

}
```

> In the next section, we will dive deeper into the process of working with `tm_dynamic` Blocks and explore practical examples.

Reviewer: ```suggestion
In the next section, we will dive deeper into the process of working with `tm_dynamic` blocks and explore practical examples.
```

# Working with `tm_dynamic` Blocks

The `tm_dynamic` block is a specialized block type that is utilized within the content block of the generate_hcl block. 

It shares similarities with Terraform dynamic blocks but it also incorporates partial evaluation of the expanded code. This allows for more flexible and dynamic code generation.

## Partial evaluation of Expanded Code

One of the key features of the `tm_dynamic` block is its ability to perform partial evaluation of the expanded code. 

This means that Terramate variables and functions within the block are evaluated, while other code elements are directly copied to the final generated code without evaluation.

## Usage of Content and Attributes in `tm_dynamic` blocks

The `tm_dynamic` block can be defined using either a content block, an attributes attribute, or a combination of both. 

The content block allows for the generation of additional sub-blocks and nesting of `tm_dynamic` blocks.

## Examples: Generating Dynamic Blocks with `tm_dynamic`

To get the more clear understanding of the concepts that we have explained above, let’s take some examples and understand those examples by looking at the generated outcome.

> ### Using the Content Block:

Reviewer: Why are we adding a colon here @Adeniyikayodee ? We don't do this in any of the other documentation sections?

}
```

In the next section, we will dive deeper into the process of working with `tm_dynamic` blocks and explore 
practical examples.

> # Working with `tm_dynamic` Blocks

Reviewer: ```suggestion
# Working with `tm_dynamic` Blocks

```

The `tm_dynamic` block is a specialized block type that is utilized within the content block of the 
`generate_hcl` block. 

It shares similarities with Terraform dynamic blocks but it also incorporates partial evaluation of the 
expanded code. This allows for more flexible and dynamic code generation.

> ## Partial Evaluation of Expanded Code

Reviewer: ```suggestion
## Partial Evaluation of Expanded Code

```

One of the key features of the `tm_dynamic` block is its ability to perform partial evaluation of the expanded 
code. 

This means that Terramate variables and functions within the block are evaluated, while other code elements are 
directly copied to the final generated code without evaluation.

> ## Usage of Content and Attributes in `tm_dynamic` Blocks

Reviewer: ```suggestion
## Usage of Content and Attributes in `tm_dynamic` Blocks

```

}
```

> ### Optional `for_each` Attribute 

Reviewer: ```suggestion
### Optional `for_each` Attribute

```

The `for_each` attribute within the `tm_dynamic` block is optional. If not defined, only a single 
block will be generated, and no iterator will be available during block generation.

> ### Optional `condition` Attribute

Reviewer: ```suggestion
### Optional `condition` Attribute

```

}
```

In this example, if the `global.values` is undefined, the `tm_dynamic` block is ignored during code generation.

# Hierarchical Code Generation

Hierarchical code generation in Terramate allows defining code generation at different levels in a project. It provides flexibility and control over the generated code. 

By specifying code generation rules at various levels, you can achieve granular control over different stacks and the resources that are associated to them.

> ## Defining Code Generation at different levels in a project

Reviewer: ```suggestion
## Defining Code Generation at different levels in a project
```

inconsistencies in upper and lower case - we discussed this the other day. Please follow a clear pattern.

## Potential impact of Code Generation on multiple or all stacks

One of the key advantages of hierarchical code generation is the ability to apply code generation rules to multiple stacks simultaneously. By defining code generation configurations at a higher level, such as the root level, you can propagate those configurations to all child modules and resources within the project. 

This allows for consistent code generation across multiple stacks, ensuring standardized infrastructure deployments.

When code generation is applied to multiple stacks, any changes or updates made to the code generation configurations at the root level will automatically reflect in all associated stacks during the next generation process. This simplifies maintenance and ensures consistent code generation across the project.

## Restrictions and Conflicts with `generate_hcl` Blocks

While hierarchical code generation provides flexibility, it's important to consider potential restrictions and conflicts that may arise when working with `generate_hcl` blocks. Some of the potential conflicts can come while:

> ### Overriding Code Generation Configurations: 

Reviewer: ```suggestion
### Overriding Code Generation Configurations: 

```

line break missing

When defining code generation configurations at different levels, it's essential to understand the order of precedence. Code generation rules at lower levels, such as module or resource-specific configurations, can override or augment the configurations defined at higher levels. 

It's crucial to carefully manage and review these configurations to avoid conflicts or unintended consequences.

> ### Conflicting Generate_hcl blocks: 

Reviewer: ```suggestion
### Conflicting Generate_hcl blocks:

```

line break missing

> If multiple `generate_hcl` blocks are defined for the same context, such as multiple generate blocks within the same module, conflicts may occur. It's important to ensure that the code generation rules within the conflicting blocks do not contradict each other or generate conflicting code. 

Reviewer: ```suggestion
If multiple `generate_hcl` blocks are defined for the same context, such as multiple generate blocks within the same module, conflicts may occur. It's important to ensure that the code generation rules within the conflicting blocks do not contradict each other or generate conflicting code. 
```

line length too long

Resolving such conflicts requires careful evaluation and adjustment of the code generation configurations.

> ### Dependency Management: 

Reviewer: ```suggestion
### Dependency Management:

```

line break missing

Hierarchical code generation relies on the dependencies between different components within the project. It's crucial to consider the dependencies and order of code generation to avoid circular dependencies or incomplete code generation. 

Understanding the relationships between modules and resources is essential to ensure the accurate generation of code.

> 

Reviewer: ```suggestion
```

> The hierarchical code generation in Terramate provides a powerful mechanism to define code generation configurations at different levels, apply them to multiple stacks, and ensure consistency across your infrastructure deployments. Understanding the hierarchy and managing potential conflicts will enable you to take full advantage of this feature and streamline your infrastructure provisioning process.

Reviewer: this line length is wayy too long

> # Conditional Code Generation

Reviewer: ```suggestion
# Conditional Code Generation

```

Conditional code generation allows you to dynamically generate code based on specified conditions. This feature enables you to control the inclusion or exclusion of certain code blocks depending on the evaluated conditions.

> ## Using the condition attribute for Conditional Code Generation:

Reviewer: ```suggestion
## Using the `condition` attribute for Conditional Code Generation:
```

}
```

In this example, if the `global.values` is undefined, the `tm_dynamic` block is ignored during code generation.

> # Hierarchical Code Generation

Reviewer: ```suggestion
# Hierarchical Code Generation

```

Hierarchical code generation in Terramate allows defining code generation at different levels in a project. 
It provides flexibility and control over the generated code. 

By specifying code generation rules at various levels, you can achieve granular control over different stacks 
and the resources that are associated to them.

> ## Defining Code Generation at Different Levels in a Project

Reviewer: ```suggestion
## Defining Code Generation at different Levels in a Project

```

Terramate supports defining code generation rules at different levels within a project structure. This allows
you to specify code generation configurations at the root level, module level, or even for individual 
resources within a module. 

This hierarchical approach enables you to tailor the generated code based on the specific requirements and
dependencies of each component.

When defining code generation at respective levels, you can utilize the `generate_hcl` block within the 
respective configuration files. Each `generate_hcl` block encapsulates the code generation rules for a specific 
context, such as a module or resource. 

By organizing code generation configurations hierarchically, you can ensure that each component generates the 
desired code in the correct context.

> ## Potential Impact of Code Generation on Multiple or all Stacks

Reviewer: ```suggestion
## Potential Impact of Code Generation on multiple or all Stacks

```

One of the key advantages of hierarchical code generation is the ability to apply code generation rules to 
multiple stacks simultaneously. By defining code generation configurations at a higher level, such as the 
root level, you can propagate those configurations to all child modules and resources within the project. 

This allows for consistent code generation across multiple stacks, ensuring standardized infrastructure 
deployments.

When code generation is applied to multiple stacks, any changes or updates made to the code generation
configurations at the root level will automatically reflect in all associated stacks during the next 
generation process. This simplifies maintenance and ensures consistent code generation across the project.

> ## Restrictions and Conflicts with `generate_hcl` Blocks

Reviewer: ```suggestion
## Restrictions and Conflicts with `generate_hcl` Blocks

```

While hierarchical code generation provides flexibility, it's important to consider potential restrictions 
and conflicts that may arise when working with `generate_hcl` blocks. 

> ### Overriding Code Generation Configurations 

Reviewer: ```suggestion
### Overriding Code Generation Configurations

```

When defining code generation configurations at different levels, it's essential to understand the order 
of precedence. Code generation rules at lower levels, such as module or resource-specific configurations, 
can override or augment the configurations defined at higher levels. 

It's crucial to carefully manage and review these configurations to avoid conflicts or unintended consequences.

> ### Conflicting `generate_hcl` Blocks 

Reviewer: ```suggestion
### Conflicting `generate_hcl` Blocks

```

If multiple `generate_hcl` blocks are defined for the same context, such as multiple generate blocks within the 
same module, conflicts may occur. It's important to ensure that the code generation rules within the 
conflicting blocks do not contradict each other or generate conflicting code. 

Resolving such conflicts requires careful evaluation and adjustment of the code generation configurations.

> ### Dependency Management 

Reviewer: ```suggestion
### Dependency Management

```

and configurations.

This concludes the subsection on conditional code generation, demonstrating how you can utilize the `condition` 
attribute to generate code selectively based on specified conditions.

# Partial Evaluation in Code Generation

## Understanding Partial Evaluation Strategy
In HCL code generation, a partial evaluation strategy is employed. This approach allows the generation of code 
with unknown references or function calls, which are then copied verbatim to the generated code. It enables a 
flexible and dynamic code generation process.

## Handling Unknown References and Function Calls
When generating HCL code, both unknown references and function calls are handled. Unknown references, such as 
Terramate references, are retained as is in the generated code. On the other hand, function calls are partially 
evaluated. 

If a function call starts with the prefix `tm_`, it is considered a Terramate function and will be evaluated. 
Function calls can have Terramate references or literals as parameters.

# Examples of Partial Evaluation in HCL Code Generation
To understand the above concepts in a better way, lets take some examples.

> ## Example 1: Mixing Terramate and Terraform References

Reviewer: ```suggestion
## Example 1: Mixing Terramate and Terraform References

```

}
```

The Terramate references `global.terramate_data and terramate.path` are evaluated, while the references to 
`var.enabled` and `local.name` are retained as is, demonstrating partial evaluation.

> ## Example 2: Partial Evaluation of Function Calls

Reviewer: ```suggestion
## Example 2: Partial Evaluation of Function Calls

```

}
```

The function call `tm_upper(global.terramate_data)` is evaluated, transforming the value of `global.
terramate_data` to uppercase. However, the function call upper `(local.name)` is retained as is since it does 
not match the Terramate function pattern.

> ## Example 3: Function Call with Terramate Reference as Parameter

Reviewer: ```suggestion
## Example 3: Function Call with Terramate Reference as Parameter

```
